File: backend/app/schemas/browser_ingestion.py
"""Browser ingestion worker queue schemas."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class BrowserIngestionStartRequest(BaseModel):
    limit: int = Field(default=100, ge=1, le=5000)
    mode: str = Field(default="both", pattern="^(transcript|comments|both)$")
    creator_filter: str | None = Field(default=None, max_length=255)
    latest_only: bool = False
    only_missing: bool = True


class BrowserIngestionJobResult(BaseModel):
    transcript_status: str | None = None
    comments_status: str | None = None
    sheets_transcript: str | None = None
    sheets_comments: str | None = None
    embedding_created: bool | None = None
    screenshot_path: str | None = None
    failure_category: str | None = None
    duration_ms: int | None = None
    current_phase: str | None = None
    retry_history: list[str] = Field(default_factory=list)
    logs: list[str] = Field(default_factory=list)


class BrowserIngestionJobRead(BaseModel):
    id: int
    run_id: int
    video_id: int
    video_url: str
    title: str
    creator_name: str
    mode: str
    status: str
    retry_count: int
    error_message: str | None
    result_json: dict[str, Any] | None
    transcript_status: str | None = None
    comments_status: str | None = None
    sheets_status: str | None = None
    embedding_status: str | None = None
    failure_category: str | None = None
    duration_seconds: float | None = None
    worker_name: str | None = None
    screenshot_path: str | None = None
    current_phase: str | None = None
    retry_history: list[str] = Field(default_factory=list)
    updated_at: datetime
    finished_at: datetime | None = None

    model_config = {"from_attributes": True}


class BrowserIngestionRunStats(BaseModel):
    jobs_total: int = 0
    queued: int = 0
    processing: int = 0
    success: int = 0
    failed: int = 0
    skipped: int = 0
    processed: int = 0
    progress_pct: float = 0.0
    success_pct: float = 0.0


class BrowserIngestionRunRead(BaseModel):
    id: int
    status: str
    mode: str
    limit_count: int | None
    creator_filter: str | None
    latest_only: bool
    only_missing: bool
    jobs_total: int
    message: str | None
    error_message: str | None
    started_at: datetime
    finished_at: datetime | None
    run_stats: BrowserIngestionRunStats
    jobs: list[BrowserIngestionJobRead] = Field(default_factory=list)


class BrowserIngestionWorkerRead(BaseModel):
    id: int
    name: str
    status: str
    current_action: str
    current_phase: str | None = None
    current_job_id: int | None
    current_video_url: str | None
    last_heartbeat_at: datetime | None
    processed_today: int = 0
    success_today: int = 0
    failed_today: int = 0
    jobs_per_min: float | None = None
    consecutive_failures: int = 0
    max_jobs_per_day: int | None = None
    daily_limit_reached: bool = False
    cooldown_until: datetime | None = None
    extension_version: str | None = None
    required_extension_version: str | None = None
    memory_mb: float | None = None
    uptime_seconds: int | None = None
    last_screenshot_path: str | None = None
    last_success_at: datetime | None = None
    restart_recommended: bool = False
    processed_per_hour: float | None = None
    health_status: str = "unknown"
    clear_local_cooldown: bool = False

    model_config = {"from_attributes": True}


class BrowserIngestionDashboard(BaseModel):
    worker: BrowserIngestionWorkerRead | None
    active_run_id: int | None
    run: BrowserIngestionRunRead | None
    catalog_videos_total: int = 0
    catalog_missing_transcript: int = 0
    catalog_missing_comments: int = 0


class BrowserIngestionStartResponse(BaseModel):
    run: BrowserIngestionRunRead
    message: str


class BrowserIngestionWorkerRegisterRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=128)


class BrowserIngestionWorkerRegisterResponse(BaseModel):
    worker_id: int
    token: str
    message: str


class BrowserIngestionHeartbeatRequest(BaseModel):
    status: str = Field(
        default="online",
        pattern="^(online|paused|incompatible_extension|cooldown|daily_limit)$",
    )
    current_action: str = "idle"
    current_phase: str | None = None
    current_job_id: int | None = None
    current_video_url: str | None = None
    processed_today: int = 0
    success_today: int = 0
    failed_today: int = 0
    jobs_per_min: float | None = None
    consecutive_failures: int = 0
    max_jobs_per_day: int | None = None
    daily_limit_reached: bool = False
    cooldown_until: str | None = None
    extension_version: str | None = None
    required_extension_version: str | None = None
    memory_mb: float | None = None
    uptime_seconds: int | None = None
    last_screenshot_path: str | None = None
    last_success_at: str | None = None
    restart_recommended: bool = False
    processed_per_hour: float | None = None


class BrowserIngestionJobClaimResponse(BaseModel):
    job: BrowserIngestionJobRead | None = None
    run_paused: bool = False


class BrowserIngestionJobCompleteRequest(BaseModel):
    result: BrowserIngestionJobResult | None = None


class BrowserIngestionJobFailRequest(BaseModel):
    error_message: str = Field(..., min_length=1, max_length=8000)
    result: BrowserIngestionJobResult | None = None
    retryable: bool = True


class BrowserIngestionJobsPage(BaseModel):
    items: list[BrowserIngestionJobRead]
    total: int
    offset: int
    limit: int
    status_filter: str | None = None

File: frontend/types/browser-ingestion.ts
/** Browser automation ingestion (local worker + Chrome extension). */

export type BrowserIngestionJobStatus =
  | "queued"
  | "processing"
  | "success"
  | "failed"
  | "skipped";

export type BrowserIngestionRunStatus =
  | "queued"
  | "running"
  | "paused"
  | "completed"
  | "failed";

export type BrowserIngestionMode = "transcript" | "comments" | "both";

export type BrowserJobStatusFilter = "all" | BrowserIngestionJobStatus;

export type FailureCategory =
  | "youtube_blocked"
  | "transcript_unavailable"
  | "comments_disabled"
  | "extension_error"
  | "browser_crash"
  | "timeout"
  | "unknown";

export type TranscriptOutcome = "ok" | "unavailable" | "failed" | "skipped";
export type CommentsOutcome = "ok" | "disabled" | "empty" | "failed" | "skipped";

export type WorkerHealthStatus =
  | "healthy"
  | "offline"
  | "incompatible_extension"
  | "daily_limit"
  | "cooldown"
  | "restart_recommended"
  | "unknown";

export interface BrowserIngestionRunStats {
  jobs_total: number;
  queued: number;
  processing: number;
  success: number;
  failed: number;
  skipped: number;
  processed: number;
  progress_pct: number;
  success_pct: number;
}

export interface BrowserIngestionJob {
  id: number;
  run_id: number;
  video_id: number;
  video_url: string;
  title: string;
  creator_name: string;
  mode: string;
  status: BrowserIngestionJobStatus;
  retry_count: number;
  error_message: string | null;
  transcript_status: string | null;
  comments_status: string | null;
  transcript_outcome: TranscriptOutcome | null;
  comments_outcome: CommentsOutcome | null;
  sheets_status: string | null;
  embedding_status: string | null;
  failure_category: FailureCategory | string | null;
  duration_seconds: number | null;
  worker_name: string | null;
  screenshot_path: string | null;
  current_phase: string | null;
  retry_history: string[];
  updated_at: string;
  finished_at: string | null;
}

export interface BrowserIngestionJobsPage {
  items: BrowserIngestionJob[];
  total: number;
  offset: number;
  limit: number;
  status_filter: string | null;
}

export interface BrowserIngestionWorker {
  id: number;
  name: string;
  status: string;
  current_action: string;
  current_phase: string | null;
  current_job_id: number | null;
  current_video_url: string | null;
  last_heartbeat_at: string | null;
  processed_today: number;
  success_today: number;
  failed_today: number;
  jobs_per_min: number | null;
  consecutive_failures: number;
  max_jobs_per_day: number | null;
  daily_limit_reached: boolean;
  cooldown_until: string | null;
  extension_version: string | null;
  required_extension_version: string | null;
  memory_mb: number | null;
  uptime_seconds: number | null;
  last_screenshot_path: string | null;
  last_success_at: string | null;
  restart_recommended: boolean;
  processed_per_hour: number | null;
  health_status: WorkerHealthStatus;
  jobs_in_run_success?: number;
  jobs_in_run_failed?: number;
  jobs_in_run_processing?: number;
}

export interface BrowserIngestionRun {
  id: number;
  status: BrowserIngestionRunStatus;
  mode: BrowserIngestionMode;
  limit_count: number | null;
  creator_filter: string | null;
  latest_only: boolean;
  only_missing: boolean;
  jobs_total: number;
  message: string | null;
  error_message: string | null;
  started_at: string;
  finished_at: string | null;
  run_stats: BrowserIngestionRunStats;
  jobs: BrowserIngestionJob[];
}

export interface BrowserIngestionDashboard {
  worker: BrowserIngestionWorker | null;
  workers?: BrowserIngestionWorker[];
  active_run_id: number | null;
  run: BrowserIngestionRun | null;
  catalog_videos_total: number;
  catalog_missing_transcript: number;
  catalog_missing_comments: number;
}

export interface BrowserIngestionStartRequest {
  limit: number;
  mode: BrowserIngestionMode;
  creator_filter?: string | null;
  latest_only: boolean;
  only_missing: boolean;
}

export interface BrowserIngestionStartResponse {
  run: BrowserIngestionRun;
  message: string;
}

export function isBrowserIngestionRunActive(
  run: BrowserIngestionRun | null | undefined,
): boolean {
  if (!run) return false;
  return run.status === "queued" || run.status === "running" || run.status === "paused";
}

export function isBrowserIngestionRunTerminal(
  run: BrowserIngestionRun | null | undefined,
): boolean {
  if (!run) return false;
  return run.status === "completed" || run.status === "failed";
}

export function isWorkerConnected(worker: BrowserIngestionWorker | null | undefined): boolean {
  if (!worker) return false;
  return worker.status !== "offline";
}

export function isWorkerHealthy(worker: BrowserIngestionWorker | null | undefined): boolean {
  if (!worker) return false;
  const dailyCapped =
    worker.max_jobs_per_day != null &&
    worker.max_jobs_per_day > 0 &&
    (worker.health_status === "daily_limit" || worker.daily_limit_reached);
  if (dailyCapped) return false;
  return worker.health_status === "healthy" || worker.status === "online";
}
